test/com/google/enterprise/gsafeed/GsafeedHelperTest.java

```diff
@@ -0,0 +1,270 @@
+        + "<!ENTITY e '&d;&d;&d;&d;&d;&d;&d;&d;&d;&d;' > "
+        + "<!ENTITY f '&e;&e;&e;&e;&e;&e;&e;&e;&e;&e;' > "
+        + "<!ENTITY g '&f;&f;&f;&f;&f;&f;&f;&f;&f;&f;' > "
+        + "<!ENTITY h '&g;&g;&g;&g;&g;&g;&g;&g;&g;&g;' > "
+        + "<!ENTITY i '&h;&h;&h;&h;&h;&h;&h;&h;&h;&h;' > "
+        + "<!ENTITY j '&i;&i;&i;&i;&i;&i;&i;&i;&i;&i;' > "
+        + "]>"
+        + "<gsafeed>"
+        + "  <header>"
+        + "    <datasource>sample</datasource>"
+        + "    <feedtype>full</feedtype>"
+        + "  </header>"
+        + "  <group>"
+        + "    <record url='http://www.example.com/hello01' "
+        + "      mimetype='text/plain'"
+        + "      last-modified='Tue, 6 Nov 2007 12:45:26 GMT'>"
+        + "      <content> &j; </content>"                            // <----
+        + "    </record>"
+        + "  </group>"
+        + "</gsafeed>";
+
+    externalEntityFeed =
+        "<?xml version='1.0' encoding='utf-8'?>"
+        + "<!DOCTYPE gsafeed PUBLIC '-//Google//DTD GSA Feeds//EN' '' [ "
+        + "<!ENTITY xxe SYSTEM '" + privateFile.toURI() + "'>"       // <----
+        + "]>"
+        + "<gsafeed>"
+        + "  <header>"
+        + "    <datasource>sample</datasource>"
+        + "    <feedtype>full</feedtype>"
+        + "  </header>"
+        + "  <group>"
+        + "    <record url='http://www.example.com/hello01' "
+        + "      mimetype='text/plain'"
+        + "      last-modified='Tue, 6 Nov 2007 12:45:26 GMT'>"
+        + "      <content> &xxe; </content>"                          // <----
+        + "    </record>"
+        + "  </group>"
+        + "</gsafeed>";
+
+    externalParameterEntityFeed =
+        "<?xml version='1.0' encoding='utf-8'?>"
+        + "<!DOCTYPE gsafeed PUBLIC '-//Google//DTD GSA Feeds//EN' '' [ "
+        + "<!ENTITY % dtd SYSTEM "
+        + "'" + evilDtd.toURI() + "'>"                               // <----
+        + "%dtd;"
+        + "%all;"
+        + "]>"
+        + "<gsafeed>"
+        + "  <header>"
+        + "    <datasource>sample</datasource>"
+        + "    <feedtype>full</feedtype>"
+        + "  </header>"
+        + "  <group>"
+        + "    <record url='http://www.example.com/hello01'"
+        + "      mimetype='text/plain'"
+        + "      last-modified='Tue, 6 Nov 2007 12:45:26 GMT'>"
+        + "      <content> &fileContents; </content>"                 // <----
+        + "    </record>"
+        + "  </group>"
+        + "</gsafeed>";
+  }
+
+  private static void writeFile(File file, String content) throws IOException {
+    FileWriter writer = new FileWriter(file);
+    try {
+      writer.write(content);
+    } finally {
+      writer.close();
+    }
+  }
+
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  @Test
+  public void testWithDtd() throws Exception {
+    Gsafeed feed = GsafeedHelper.unmarshalWithDtd(
+        new ByteArrayInputStream(validFeed.getBytes("UTF-8")));
```

Comment:
Use a Charset named UTF_8, as in SendGsaFeed. I'm also tempted by an asStream helper method or something here, or maybe this suggests that GsafeedHelper itself should have some overloads (it depends on whether strings are viewed as test-specific or a general use case).